### PR TITLE
[Feature] Add Precise BN

### PR DESCRIPTION
**Pull request by Ezra-Yu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add PreciseBN Hook.

## Modification

1. Add PreciseBN Hook and corresponding unit tests.
2. Add resnet example PreciseBN config.

refer to [mmaction](https://github.com/open-mmlab/mmaction2/pull/501) and [pycls](https://github.com/facebookresearch/pycls/blob/f8cd962737e33ce9e19b3083a33551da95c2d9c0/pycls/core/net.py#L25).

## BC-breaking (Optional)

No!

## Use cases (Optional)

add in config

```
custom_hooks = [
    dict(
        type='PreciseBNHook',
        num_items=8192,
        interval=1,
        priority='ABOVE_NORMAL')
]
```

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by mzr1996** [comment]
Please put hooks into `mmcls/runner/hooks/precise_bn.py`, refers to [mmcv](https://github.com/open-mmlab/mmcv/tree/master/mmcv/runner/hooks).

File: mmcls/runner/hooks/precise_bn.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            raise TypeError(
+                'dataloaders must be a list of Pytorch Dataloaders ,but got',
+                f' {type(dataloaders[0])}')
+        self.dataloaders = dataloaders
+        self.interval = interval
+        self.num_items = num_items
+
+    def after_train_epoch(self, runner):
+        if self.every_n_epochs(runner, self.interval):
+            # sleep to avoid possible deadlock
```
Comment (mzr1996) [inline mmcls/runner/hooks/precise_bn.py:164]: possible deadlock?

Comment (Ezra-Yu) [reply mmcls/runner/hooks/precise_bn.py:164]: I refer to the code of [mmaction2](https://github.com/open-mmlab/mmaction2/blob/075464c9157a20a33ccecf11ed3a532ecfbb3b78/mmaction/utils/precise_bn.py#L144). 

There should be no risk of deadlock here. 
Both two "sleep" lines can be deleted here.

File: tests/test_runtime/test_preciseBN_hook.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+            device_ids=[torch.cuda.current_device()],
+            broadcast_buffers=False,
+            find_unused_parameters=True)
+        runner = EpochBasedRunner(
+            model=model,
+            batch_processor=None,
+            optimizer=optimizer,
+            logger=logger)
+        runner.register_hook(precise_bn_hook)
+        runner.run(loaders, [('train', 1)], 1)
```
Comment (mzr1996) [inline tests/test_runtime/test_preciseBN_hook.py:224]: Please set `max_epochs` when instantiating the runner instead of passing it in `run`.

**Review by mzr1996 — CHANGES_REQUESTED** [review]

File: tests/test_runtime/test_preciseBN_hook.py (hunk 2 above)
Comment (Ezra-Yu) [reply tests/test_runtime/test_preciseBN_hook.py]: done.

File: mmcls/core/hook/precise_bn_hook.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    for i, bn in enumerate(bn_layers):
+        bn.running_mean = running_means[i]
+        bn.running_var = running_vars[i]
+        bn.momentum = momentums[i]
+
+
+@HOOKS.register_module()
+class PreciseBNHook(Hook):
+    """Precise BN hook.
+
```
Comment (ZwwWayne) [inline mmcls/core/hook/precise_bn_hook.py:153]: May indicate its recommended priority when it is used. For example, its priority should be "xxx" and should be executed before xxx hook after xxx hook

File: configs/resnet/resnet50_8xb32-coslr-preciseBN_in1k.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+_base_ = 'resnet50_8xb32-coslr_in1k.py'
+
+# precise BN hook
+custom_hooks = [
+    dict(
+        type='PreciseBNHook',
+        num_items=8192,
+        interval=1,
+        priority='ABOVE_NORMAL')
```
Comment (ZwwWayne) [inline configs/resnet/resnet50_8xb32-coslr-preciseBN_in1k.py:9]: May add comments here to indicate why its priority is ABOVE_NORMAL.

File: mmcls/core/hook/precise_bn_hook.py (hunk 3 above)
Comment (Ezra-Yu) [reply mmcls/core/hook/precise_bn_hook.py:135]: done.

File: configs/resnet/resnet50_8xb32-coslr-preciseBN_in1k.py (hunk 4 above)
Comment (Ezra-Yu) [reply configs/resnet/resnet50_8xb32-coslr-preciseBN_in1k.py:11]: done.

**Review by mzr1996 — APPROVED** [review]
LGTM

**Pull request merged**